### PR TITLE
Improve speed of RoadPathFinder

### PR DESCRIPTION
**Pull request by Flamefire**
I discovered 2 main hotspots in the roadpatfinder:
- Calling GetNeighbour due to multiple, duplicate checks and pointer dereferences -> Introduce GetRoutes to get all routes at once
- `Flag::GetPunishmentPoints` iterated over an array of 8 wares, that is usually empty and can be skipped. Using a fixed capacity vector avoids that but breaks replays

As replays were already broken I decided to introduce the compression I mentioned in  #1383 which shrinks savegames by 90% and replays starting at GF 0 by 25-50% which IMO is worth it, especially as the tests contain replays so the repo size gets reduced

- [x] Based on #1391

### CODE REVIEW

File: libs/s25main/Replay.cpp [hunk 1]
```diff
@@ -32,28 +33,76 @@ uint16_t Replay::GetVersion() const
...
+        if(data.size() >= uncompressedSize)
+            return true;
+        compressedReplay.WriteUnsignedInt(uncompressedSize);
+        compressedReplay.WriteUnsignedInt(data.size());
+        compressedReplay.WriteRawData(data.data(), data.size());
+
+        // All done. Replace uncompressed replay
+        compressedReplay.Close();
+        file.Close();
+        boost::filesystem::rename(tmpReplayFile.filePath, filepath_);
```
Comment (Flow86) [inline libs/s25main/Replay.cpp:99]: does this overwrite existing files?

File: libs/s25main/gameTypes/CompressedData.cpp [hunk 2]
```diff
@@ -33,61 +34,82 @@ bool CompressedData::DecompressToFile(const boost::filesystem::path& filePath, u
...
-    unsigned compressedLen = data.size();
-    int err = BZ2_bzBuffToBuffCompress(&data[0], &compressedLen, uncompressedData.get(), length, 9, 0, 250);
-    if(err != BZ_OK)
+    try
+    {
+        data = compress(uncompressedData);
+    } catch(const std::runtime_error& err)
     {
-        LOG.write("FATAL ERROR: BZ2_bzBuffToBuffCompress failed with error: %d\n") % err;
+        LOG.write("FATAL ERROR: 1%\n") % err.what();
```
Comment (Flow86) [inline libs/s25main/gameTypes/CompressedData.cpp:76]: isnt there a % missing before the 1?

File: libs/s25main/Replay.cpp (hunk 1 above)
Comment (Flamefire) [reply libs/s25main/Replay.cpp:99]: Yes and it should as it is meant to overwrite the just created replay: https://www.boost.org/doc/libs/1_45_0/libs/filesystem/v3/doc/reference.html#rename

File: libs/s25main/gameTypes/CompressedData.cpp (hunk 2 above)
Comment (Flamefire) [reply libs/s25main/gameTypes/CompressedData.cpp]: Yes, good catch!
It is practically impossible to trigger this codepath though, so can't be tested: http://linux.math.tifr.res.in/manuals/html/manual_3.html#SEC36

**Review by Flow86 — CHANGES_REQUESTED** [review]
conflict, rebase necessary

**Pull request merged**